File: python/sunlight/ids.py
"""Locked ID prefixes and nanoid generation (spec §2)."""

from __future__ import annotations

import secrets

ID_SUFFIX_LENGTH = 21
ID_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-"
ID_PREFIXES = ("sll", "slk", "sls", "slq", "sle", "slr", "sla")


def is_valid_id(value, prefix: str | None = None) -> bool:
    if not isinstance(value, str):
        return False
    sep = value.find("_")
    if sep <= 0:
        return False
    p, suffix = value[:sep], value[sep + 1 :]
    if p not in ID_PREFIXES or (prefix is not None and p != prefix):
        return False
    return len(suffix) == ID_SUFFIX_LENGTH and all(c in ID_ALPHABET for c in suffix)


def new_id(prefix: str, rand=secrets.token_bytes) -> str:
    """Cryptographic nanoid over the 64-symbol alphabet (divides 256 evenly)."""
    if prefix not in ID_PREFIXES:
        raise ValueError(f"unknown id prefix {prefix!r}")
    out = []
    while len(out) < ID_SUFFIX_LENGTH:
        for b in rand(ID_SUFFIX_LENGTH):
            if len(out) < ID_SUFFIX_LENGTH:
                out.append(ID_ALPHABET[b & 63])
    return f"{prefix}_{''.join(out)}"


def random_hex(n: int) -> str:
    return secrets.token_bytes(n).hex()

File: src/ids.ts
/**
 * Locked ID prefixes and nanoid generation (spec §2).
 *
 * Every ID is `<prefix>` + `_` + exactly 21 characters from the 64-symbol
 * alphabet [A-Za-z0-9_-].  Generation uses cryptographic rejection sampling
 * (nanoid algorithm).  Prefixes are wire constraints, not display
 * conventions.
 */

import { randomBytes } from "node:crypto";

export const ID_SUFFIX_LENGTH = 21;
export const ID_ALPHABET =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-";

export const ID_PREFIXES = [
  "sll", // LedgerId
  "slk", // KeyId
  "sls", // StatementId
  "slq", // RequestId / CommandId
  "sle", // EntryId
  "slr", // RunId
  "sla", // ActionId
] as const;

export type IdPrefix = (typeof ID_PREFIXES)[number];

const PREFIX_SET = new Set<string>(ID_PREFIXES);

export function isValidId(value: unknown, prefix?: IdPrefix): value is string {
  if (typeof value !== "string") return false;
  const sep = value.indexOf("_");
  if (sep <= 0) return false;
  const p = value.slice(0, sep);
  if (!PREFIX_SET.has(p)) return false;
  if (prefix !== undefined && p !== prefix) return false;
  const suffix = value.slice(sep + 1);
  if (suffix.length !== ID_SUFFIX_LENGTH) return false;
  for (const ch of suffix) {
    if (ID_ALPHABET.indexOf(ch) < 0) return false;
  }
  return true;
}

/**
 * Cryptographic nanoid with rejection sampling. The 64-symbol alphabet divides
 * 256 evenly, so the mask path never discards; the sampler is still the
 * standard rejection loop and remains correct if the alphabet ever changes.
 */
export function newId(prefix: IdPrefix, random: (n: number) => Buffer = randomBytes): string {
  const alphabet = ID_ALPHABET;
  const size = alphabet.length;
  let mask = 1;
  while (mask < size - 1) mask = (mask << 1) | 1;
  const bound = Math.floor(256 / size) * size;
  const out = new Array<string>(ID_SUFFIX_LENGTH);
  let filled = 0;
  while (filled < ID_SUFFIX_LENGTH) {
    const bytes = random(ID_SUFFIX_LENGTH);
    for (let i = 0; i < bytes.length && filled < ID_SUFFIX_LENGTH; i++) {
      const b = bytes[i]! & mask;
      if (b < size && bytes[i]! < bound) {
        out[filled++] = alphabet[b]!;
      }
    }
  }
  return `${prefix}_${out.join("")}`;
}

/** Lowercase hex of `n` random bytes (head.get nonces etc.). */
export function randomHex(n: number, random: (n: number) => Buffer = randomBytes): string {
  return random(n).toString("hex");
}
